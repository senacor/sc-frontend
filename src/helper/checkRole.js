import ROLES from './roles';

export const hasRole = (role, userroles) => {
  return userroles.includes(role);
};

export const isEmployee = (userroles = []) => {
  return userroles.includes('PR_Mitarbeiter');
};

export const isSupervisor = userroles => {
  return userroles.includes(ROLES.PR_CST_LEITER);
};
