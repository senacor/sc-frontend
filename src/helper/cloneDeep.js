const cloneDeep = require('lodash/cloneDeep');

export default cloneDeep;
