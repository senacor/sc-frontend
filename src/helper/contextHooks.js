import { useContext } from 'react';
import { ErrorContext, InfoContext, UserinfoContext } from '../components/App';
import ROLES from './roles';

export const useInfoContext = () => {
  const infoContext = useContext(InfoContext.context);
  return {
    context: infoContext,
    msg: msg => {
      infoContext.setValue({
        hasInfos: true,
        messageId: msg
      });
    },
    hide: () => {
      infoContext.setValue({
        hasInfos: false,
        messageId: ''
      });
    }
  };
};

export const useUserinfoContext = () => {
  const userInfoContext = useContext(UserinfoContext.context);
  const { userroles, userinfo } = userInfoContext.value;
  return {
    context: userInfoContext,
    isSupervisorInPr: pr => {
      return userinfo.userId === pr.supervisor.id;
    },
    isReviewerInPr: pr => {
      return userinfo.userId === pr.reviewer.id;
    },
    isOwnerInPr: pr => {
      return pr.employee.id === userinfo.userId;
    },
    hasRoleAdmin: () => {
      return userroles[0] === 'ADMIN';
    },
    hasRoleHr: () => {
      return userroles.includes(ROLES.PERSONAL_DEV);
    },
    hasRoleEmployee: () => {
      return userroles.includes(ROLES.DEVELOPER);
    },
    hasRoleSupervisor: () => {
      return userroles.includes(ROLES.SUPERVISOR);
    }
  };
};

export const useErrorContext = () => {
  const errorContext = useContext(ErrorContext.context);
  return {
    context: errorContext,
    showGeneral: () => {
      errorContext.setValue({
        hasErrors: true,
        messageId: 'message.error'
      });
    },
    hide: () => {
      errorContext.setValue({ hasErrors: false, messageId: '' });
    },
    hasError: error => {
      return errorContext.value.errors && errorContext.value.errors[error];
    },
    show: (msgId, errors) => {
      errorContext.setValue(
        errors
          ? { hasErrors: true, messageId: msgId, errors: errors }
          : { hasErrors: true, messageId: msgId }
      );
    }
  };
};
