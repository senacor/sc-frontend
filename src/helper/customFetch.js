import store from '../store';
import { LOGIN_UNAUTHORIZED } from './dispatchTypes';
/**
 * supported usage:
 *   fetch('URL', configObj);
 *
 *   where configObj is something like:
 *   {
 *     method,
 *     headers,
 *     body
 *   }
 */
export default function customFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`,
    'Content-Type': 'application/json'
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}

export function fileFetch(url, config = {}) {
  let access_token = localStorage.getItem('access_token');

  let authenticationConfig = Object.assign({}, config);
  authenticationConfig.headers = Object.assign({}, config.headers, {
    Authorization: `Bearer ${access_token}`
  });
  return fetch(url, authenticationConfig).then(response => {
    if (response.status === 401) {
      store.dispatch({
        type: LOGIN_UNAUTHORIZED,
        payload: response.status
      });
    }

    return response;
  });
}
