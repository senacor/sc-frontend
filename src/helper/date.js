import moment from 'moment';
import 'moment-timezone';

export const FRONTEND_DATE_FORMAT = 'DD.MM.YYYY';
export const FRONTEND_LOCALE_DATE_TIME_FORMAT = 'YYYY-MM-DD';

export const formatDateForFrontend = input => {
  if (input) {
    let date = moment.isMoment(input) ? input : moment(input);
    if (date.isValid()) {
      return date.format(FRONTEND_DATE_FORMAT);
    }
  }
  return null;
};

// Converting LocaleDateTime for frontend
export const formatLocaleDateTime = (input, format) => {
  if (input) {
    let modifiedInput = [...input];
    modifiedInput[1] = modifiedInput[1] - 1;
    let date = moment.isMoment(modifiedInput)
      ? modifiedInput
      : moment.utc(modifiedInput);
    if (date.isValid()) {
      return date.tz('Europe/Berlin').format(format);
    }
  }
  return null;
};

// Converting date to LocaleDateTime for backend
export const convertLocalDateTime = date => {
  const convertedDate = moment
    .tz(`${date.target.value}`, 'Europe/Berlin')
    .utc()
    .format('YYYY-MM-DDTHH:mm');
  if (convertedDate === 'Invalid date') {
    return null;
  } else return convertedDate;
};

export const isValidDate = date => {
  const isValidDate = moment(
    date,
    FRONTEND_LOCALE_DATE_TIME_FORMAT,
    true
  ).isValid();
  return isValidDate;
};

export const getReadableDate = date => {
  const minutes = date[4] > 9 ? date[4].toString() : `0${date[4]}`;
  return `${date[2]}.${date[1]}.${date[0]}, ${date[3]}:${minutes}`;
};
