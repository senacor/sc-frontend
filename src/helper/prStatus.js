export const prStatusEnum = {
  RELEASED_SHEET_REVIEWER: 'FILLED_SHEET_REVIEWER',
  RELEASED_SHEET_EMPLOYEE: 'FILLED_SHEET_EMPLOYEE',
  REQUESTED_DATE: 'REQUESTED_DATE',
  FIXED_DATE: 'ALL_DATES_ACCEPTED',
  FINALIZED_REVIEWER: 'MODIFICATIONS_ACCEPTED_REVIEWER',
  FINALIZED_EMPLOYEE: 'MODIFICATIONS_ACCEPTED_EMPLOYEE',
  ARCHIVED_HR: 'PR_COMPLETED'
};
