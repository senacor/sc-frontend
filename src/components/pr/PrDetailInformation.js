import React, { useContext } from 'react';
import moment from 'moment-timezone';
import { injectIntl } from 'react-intl';
import { withStyles } from '@material-ui/core';
import Avatar from '@material-ui/core/Avatar/Avatar';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import List from '@material-ui/core/List/List';
import ListItem from '@material-ui/core/ListItem/ListItem';

import getDisplayName from '../../helper/getDisplayName';
import { formatDateForFrontend } from '../../helper/date';
import Grid from '@material-ui/core/Grid';
import BackToTableButton from './BackToTableButton';
import ShowReviewer from './ShowReviewer';
import { prStatusEnum } from '../../helper/prStatus';
import PrHistory from './PrHistory';
import { UserinfoContext } from '../App';

const styles = theme => ({
  root: {
    margin: 3 * theme.spacing.unit
  },
  avatarContainer: {
    flex: '0 0 auto',
    margin: 'auto 0',
    marginRight: theme.spacing.unit
  },
  avatar: {
    backgroundColor: theme.palette.primary[500]
  },
  heading: {
    flex: '1 1 auto'
  },
  list: {
    margin: 0,
    padding: 0
  },
  backBtnContainer: {
    display: 'flex',
    alignItems: 'center'
  },
  backBtn: {
    backgroundColor: theme.palette.primary[400],
    color: theme.palette.secondary.white,
    width: 150,
    marginTop: theme.spacing.unit,
    marginBottom: theme.spacing.unit
  }
});

const PrDetailInformation = ({ classes, pr, meeting, intl }) => {
  const { userinfo } = useContext(UserinfoContext.context).value;
  const { username } = userinfo;
  if (!pr) {
    return null;
  }

  const { firstName, lastName } = pr.employee;
  let termin = intl.formatMessage({
    id: 'prdetailinformation.notarranged'
  });

  if (meeting) {
    const meetingDate = meeting.start
      ? moment(meeting.start)
          .local()
          .format('DD.MM.YYYY')
      : null;
    const meetingDay = formatDateForFrontend(pr.meetingDay);

    if (meeting.status === 'ACCEPTED') {
      termin = meetingDate;
    }
    if (meeting.status === 'NO_ANSWER') {
      termin =
        meetingDate +
        ` ${intl.formatMessage({
          id: 'prdetailinformation.notconfirmed'
        })}`;
    }
    if (meeting.status === 'DECLINED') {
      termin =
        meetingDate +
        ` ${intl.formatMessage({
          id: 'prdetailinformation.cancelled'
        })}`;
    }
    if (
      pr.statusSet.includes(prStatusEnum.FINALIZED_REVIEWER) &&
      meeting.status !== 'ACCEPTED'
    ) {
      termin =
        meetingDay +
        ` ${intl.formatMessage({
          id: 'prdetailinformation.arrangedoffportal'
        })}`;
    }
  }

  const competence = intl.formatMessage({
    id: `COMPETENCE_${pr.competence}`
  });

  const subheader = `${intl.formatMessage({
    id: 'prdetailinformation.duedate'
  })} ${formatDateForFrontend(pr.deadline)}, ${intl.formatMessage({
    id: 'prdetailinformation.cc'
  })} ${competence}, ${intl.formatMessage({
    id: 'prdetailinformation.occasion'
  })} ${intl.formatMessage({
    id: `${pr.occasion}`
  })}, ${intl.formatMessage({
    id: 'prdetailinformation.termin'
  })} ${termin}`;

  const supervisorHeader = `${intl.formatMessage({
    id: 'prdetailinformation.supervisor'
  })} ${getDisplayName(pr.supervisor)}`;

  return (
    <div className={classes.root}>
      <ExpansionPanel>
        <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
          <Grid container alignItems="center" justify="center">
            <Grid item md={1} xs={3}>
              <div className={classes.avatarContainer}>
                <Avatar className={classes.avatar}>
                  {`${firstName.charAt(0)}${lastName.charAt(0)}`}
                </Avatar>
              </div>
            </Grid>
            <Grid item md={9} xs={9}>
              <div className={classes.heading}>
                <Typography variant={'body2'}>
                  {getDisplayName(pr.employee)}
                </Typography>
                <Typography variant={'body2'} color={'textSecondary'}>
                  {subheader}
                </Typography>
                <List className={classes.list}>
                  <ListItem className={classes.list}>
                    <Typography variant={'body2'} color={'textSecondary'}>
                      {supervisorHeader}
                    </Typography>
                    <ShowReviewer
                      pr={pr}
                      prefix={`, ${intl.formatMessage({
                        id: 'prdetailinformation.reviewer'
                      })} `}
                      username={username}
                    />
                  </ListItem>
                </List>
              </div>
            </Grid>
            <Grid item md={2} xs={12} style={{ flexBasis: '0' }}>
              <div className={classes.backBtnContainer}>
                <BackToTableButton pr={pr} classes={classes.backBtn} />
              </div>
            </Grid>
          </Grid>
        </ExpansionPanelSummary>
        <PrHistory employeeId={pr.employee.id} />
      </ExpansionPanel>
    </div>
  );
};

export default injectIntl(withStyles(styles)(PrDetailInformation));
