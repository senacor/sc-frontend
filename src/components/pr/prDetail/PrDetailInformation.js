import React, { Component } from 'react';
import { withStyles } from '@material-ui/core';
import Avatar from '@material-ui/core/Avatar/Avatar';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';

import Grid from '@material-ui/core/Grid';

import { translateContent } from '../../translate/Translate';
import getDisplayName from '../../../helper/getDisplayName';
import { formatDateForFrontend } from '../../../helper/date';
import { connect } from 'react-redux';
import List from '@material-ui/core/List/List';
import ListItem from '@material-ui/core/ListItem/ListItem';
import BackToTableButton from './BackToTableButton';
import { getUserPrincipalName } from '../../../reducers/selector';
import ShowReviewer from './ShowReviewer';
import { prStatusEnum } from '../../../helper/prStatus';
import moment from 'moment-timezone';
import PrHistory from './PrHistory';

const styles = theme => ({
  root: {
    marginBottom: 2 * theme.spacing.unit
  },
  avatarContainer: {
    flex: '0 0 auto',
    marginRight: 2 * theme.spacing.unit
  },
  avatar: {
    backgroundColor: theme.palette.primary[500]
  },
  heading: {
    flex: '1 1 auto'
  },
  list: {
    margin: '0px',
    padding: '0px'
  },
  buttonDesktop: {
    position: 'relative',
    backgroundColor: theme.palette.primary['400'],
    color: '#FFF',
    marginBottom: '2%',
    marginTop: '12%'
  }
});

export class PrDetailInformation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false
    };
  }

  render() {
    const { classes, pr, username, meeting } = this.props;

    if (!pr) {
      return null;
    }

    const { firstName, lastName } = pr.employee;
    let termin = 'noch nicht vereinbart';

    if (meeting) {
      const meetingDate = meeting.start
        ? moment(meeting.start)
            .local()
            .format('DD.MM.YYYY')
        : null;
      const meetingDay = formatDateForFrontend(pr.meetingDay);

      if (meeting.status === 'ACCEPTED') {
        termin = meetingDate;
      }
      if (meeting.status === 'NO_ANSWER') {
        termin = meetingDate + ' (noch nicht bestätigt)';
      }
      if (meeting.status === 'DECLINED') {
        termin = meetingDate + ' (abgesagt)';
      }
      if (
        pr.statuses.includes(prStatusEnum.FINALIZED_REVIEWER) &&
        meeting.status !== 'ACCEPTED'
      ) {
        termin = meetingDay + ' (außerhalb des Portals vereinbart)';
      }
    }
    const competence = translateContent('COMPETENCE_' + pr.competence);

    const subheader = `Fälligkeit: ${formatDateForFrontend(
      pr.deadline
    )}, ACD_BOOT, ${competence}, ${translateContent(
      pr.occasion
    )}, Termin: ${termin}`;

    const supervisorHeader = `Vorgesetzter: ${getDisplayName(pr.supervisor)}`;

    return (
      <div className={classes.root}>
        <ExpansionPanel>
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            <div className={classes.avatarContainer}>
              <Avatar className={classes.avatar}>
                {`${firstName.charAt(0)}${lastName.charAt(0)}`}
              </Avatar>
            </div>
            <div className={classes.heading}>
              <Typography variant={'body2'}>
                {getDisplayName(pr.employee)}
              </Typography>
              <Typography variant={'body2'} color={'textSecondary'}>
                {subheader}
              </Typography>
              <List className={classes.list}>
                <ListItem className={classes.list}>
                  <Typography variant={'body2'} color={'textSecondary'}>
                    {supervisorHeader}
                  </Typography>
                  <ShowReviewer
                    pr={pr}
                    prefix={', Beurteiler: '}
                    username={username}
                  />
                </ListItem>
              </List>
            </div>
            <div>
              <BackToTableButton pr={pr} classes={classes.buttonDesktop} />
            </div>
          </ExpansionPanelSummary>
          <ExpansionPanelDetails className={classes.details}>
            <Grid container spacing={24}>
              <Grid item xs={4}>
                <PrHistory employeeId={pr.employee.id} />
              </Grid>
              <Grid item xs={4} />
            </Grid>
          </ExpansionPanelDetails>
        </ExpansionPanel>
      </div>
    );
  }
}

export const StyledComponent = withStyles(styles)(PrDetailInformation);

export default connect(
  state => ({
    username: getUserPrincipalName(state)
  }),
  {}
)(StyledComponent);
