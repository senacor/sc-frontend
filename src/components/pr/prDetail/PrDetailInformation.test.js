import React from 'react';
import { StyledComponent as PrDetailInformation } from './PrDetailInformation';
import { createShallow } from '@material-ui/core/test-utils';
import { prStatusEnum } from '../../../helper/prStatus';

describe('PrDetailInformation Component', () => {
  const pr = {
    employee: {
      id: 501,
      login: 'mmitarbeiterin',
      firstName: 'Michaela',
      lastName: 'Mitarbeiterin'
    },
    supervisor: {
      id: 54,
      login: 'mbock',
      firstName: 'Manuela',
      lastName: 'Bock'
    },
    reviewer: {
      id: 503,
      login: 'mmitarbeiter',
      firstName: 'Martin',
      lastName: 'Mitarbeiter'
    },
    deadline: '2018-12-06',
    meeting: { start: '2018-12-04' },
    competence: 'DEVELOPMENT',
    occasion: 'ON_DEMAND',
    statuses: [prStatusEnum.FIXED_DATE]
  };

  let shallow = createShallow({ dive: true });

  it('should match snapshot', () => {
    const component = shallow(
      <PrDetailInformation pr={pr} username={'mmitarbeiterin'} />
    );
    expect(component).toMatchSnapshot();
  });
});
