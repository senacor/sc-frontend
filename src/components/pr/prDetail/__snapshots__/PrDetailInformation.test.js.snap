// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrDetailInformation Component should match snapshot 1`] = `
<div
  className="PrDetailInformation-root-1"
>
  <WithStyles(ExpansionPanel)>
    <WithStyles(ExpansionPanelSummary)
      className="PrDetailInformation-summary-6"
      expandIcon={<pure(ExpandMoreIcon) />}
    >
      <div
        className="PrDetailInformation-avatarContainer-3"
      >
        <WithStyles(Avatar)
          className="PrDetailInformation-avatar-4"
        >
          MM
        </WithStyles(Avatar)>
      </div>
      <div
        className="PrDetailInformation-heading-5"
      >
        <WithStyles(Typography)
          variant="body2"
        >
          Michaela Mitarbeiterin
        </WithStyles(Typography)>
        <WithStyles(Typography)
          color="textSecondary"
          variant="body2"
        >
          FÃ¤lligkeit: 06.12.2018, ACD_BOOT, DEV, Auf Nachfrage, Termin: 2018-12-04
        </WithStyles(Typography)>
        <WithStyles(Typography)
          color="textSecondary"
          variant="body2"
        >
          Vorgesetzter: Manuela Bock, Beurteiler: Martin Mitarbeiter
        </WithStyles(Typography)>
      </div>
    </WithStyles(ExpansionPanelSummary)>
    <WithStyles(ExpansionPanelDetails)>
      <WithStyles(Grid)
        container={true}
        spacing={24}
      >
        <WithStyles(Grid)
          item={true}
          xs={4}
        >
          <PrHistory />
        </WithStyles(Grid)>
        <WithStyles(Grid)
          item={true}
          xs={4}
        />
      </WithStyles(Grid)>
    </WithStyles(ExpansionPanelDetails)>
  </WithStyles(ExpansionPanel)>
</div>
`;
